File: strings/minimum-window-substring/minimum-window-substring.js
/**
 * @param {string} s
 * @param {string} t
 * @return {string}
 */
function minWindow(s, t) {
  const n = s.length;

  const freq = {};

  for (let i = 0; i < t.length; i++) {
    const c = t[i];
    if (freq[c] === undefined) {
      freq[c] = 0;
    }
    freq[c]++;
  }

  const unique = Object.keys(freq).length;

  let left = 0;
  let right = 0;
  let count = 0;

  let minLeft = -1;
  let minRight = -1;
  let minLen = Number.MAX_SAFE_INTEGER;

  const window = {};

  while (right < n) {
    const cur = s[right];

    if (freq[cur] !== undefined) {
      if (window[cur] === undefined) {
        window[cur] = 0;
      }

      window[cur]++;

      if (freq[cur] === window[cur]) {
        count++;
      }
    }

    while (left <= right && unique === count) {
      let len = right - left + 1;

      if (minLeft === -1 || len < minLen) {
        minLen = len;
        minLeft = left;
        minRight = right;
      }

      const c = s[left];

      if (freq[c] !== undefined) {
        window[c]--;

        if (window[c] < freq[c]) {
          count--;
        }
      }
      left++;
    }
    right++;
  }

  if (minLeft === -1) {
    return '';
  }

  return s.substring(minLeft, minRight + 1);
}

File: strings/minimum-window-substring/minimum-window-substring.java
class Solution {
  public String minWindow(String s, String t) {
    int n = s.length();

    Map<Character, Integer> freq = new HashMap<>();

    for (char c : t.toCharArray()) {
      freq.put(c, freq.getOrDefault(c, 0) + 1);
    }

    int unique = freq.size();

    int left = 0;
    int right = 0;
    int count = 0;

    int minLeft = -1;
    int minRight = -1;
    int minLen = Integer.MAX_VALUE;

    Map<Character, Integer> window = new HashMap<>();

    while (right < n) {
      char cur = s.charAt(right);

      if (freq.containsKey(cur)) {
        window.put(cur, window.getOrDefault(cur, 0) + 1);

        if (freq.get(cur).equals(window.get(cur))) {
          count++;
        }
      }

      while (left <= right && unique == count) {
        int len = right - left + 1;

        if (minLeft == -1 || len < minLen) {
          minLen = len;
          minLeft = left;
          minRight = right;
        }

        char c = s.charAt(left);

        if (freq.containsKey(c)) {
          window.put(c, window.get(c) - 1);

          if (window.get(c) < freq.get(c)) {
            count--;
          }
        }
        left++;
      }
      right++;
    }

    if (minLeft == -1) {
      return "";
    }

    return s.substring(minLeft, minRight + 1);
  }
}
